README.md
# ECIES - Elliptical Curve Integrated Encryption System

## Description

This library implements Elliptical Curve Integrated Encryption System (ECIES), as specified by [SEC 1: Elliptic Curve Cryptography, Version 2.0](http://www.secg.org/sec1-v2.pdf).

ECIES is a public-key encryption scheme based on ECC. It is designed to be semantically secure in the presence of an adversary capable of launching chosen-plaintext and chosen-ciphertext attacks.

ECIES can be used to encrypt messages to bitcoin addresses with public keys published on the blockchain, and subsequently to decrypt messages by the holders of the address's private key.

## Installation

This library is distributed as a gem named [ecies](https://rubygems.org/gems/ecies) at RubyGems.org.  To install it, run:

    gem install ecies

## Usage

First, require the gem:

```ruby
require 'ecies'
```

Intitlialize a key and a `Crypt` object.

```ruby
key = OpenSSL::PKey::EC.generate('secp256k1')
crypt = ECIES::Crypt.new
```

Next, we'll encrypt a message. Although in this example our key contains both the private and public components, you only need the key to contain the public component to encrypt a message.

```ruby
encrypted = crypt.encrypt(key, 'secret message')
```

Finally, decrypt the message. In order to decrypt, the key must contain the private component.

```ruby
crypt.decrypt(key, encrypted) # => "secret message"
```

### Encrypting a message to a Bitcoin address

Bitcoin P2PKH addresses themselves contain only *hashes* of public keys (hence the name, pay-to-public-key-hash). However, any time a P2PKH output is spent, the public key associated with the address is published on the blockchain in the transaction's scriptSig. This allows you to encrypt a message to any bitcoin address that has sent a transaction (or published its public key in other ways). To demonstrate this, we'll encrypt a message to Satoshi's public key from Bitcoin's genesis block:

```ruby
public_key = ECIES::Crypt.public_key_from_hex(
    "04678afdb0fe5548271967f1a67130b7105cd6a828e03909a67962e0ea1f61deb"\
    "649f6bc3f4cef38c4f35504e51ec112de5c384df7ba0b8d578a4c702b6bf11d5f")
encrypted = ECIES::Crypt.new.encrypt(public_key, 'secret message')
```

To decrypt this message, Satoshi would follow these steps:

```ruby
private_key = OpenSSL::PKey::EC.new("<PEM/DER encoded private key for genesis block>")
ECIES::Crypt.new.decrypt(private_key, encrypted) # => "secret message"
```

### Default parameters

By default, when constructing a new `ECIES::Crypt` object, it will use the following parameters for ECIES:

 - KDF: ANSI-X9.63-KDF with SHA256
 - MAC: HMAC-SHA-256-128
 - Cipher: AES-256-CTR
 - EC Group: secp256k1

These defaults work well for encrypting messages to bitcoin keys. This library also supports alternate algorithms as described in the below 'Compatibility' section. In order to utilize these other algorithms, initialize an `ECIES::Crypt` object with alternate parameters (see the `ECIES::Crypt.new` documentation for details). The `Crypt` object must be initialized with the same parameters when encrypting and decrypting messages.

## Compatibility

The sec1-v2 document allows for many combinations of various algorithms for ECIES. This library only supports a subset of the allowable algorithms:
  - Key Derivation Functions
    - Supported:
      - ANSI-X9.63-KDF
    - Not supported:
      - IKEv2-KDF
      - TLS-KDF
      - NIST-800-56-Concatenation-KDF
  - Hash Functions
    - Supported:
      - SHA-224
      - SHA-256
      - SHA-384
      - SHA-512
    - Not supported:
      - SHA-1
  - MAC Schemes
    - Supported:
      - HMAC-SHA-224-112
      - HMAC-SHA-224-224
      - HMAC-SHA-256-128
      - HMAC-SHA-256-256
      - HMAC-SHA-384-192
      - HMAC-SHA-384-384 (I believe sec1-v2 has a typo here, they state "HMAC-SHA-384-284". 284 bits would be 35.5 bytes, which is nonsensical)
      - HMAC-SHA-512-256
      - HMAC-SHA-512-512
    - Not supported:
      - HMAC-SHA-1-160
      - HMAC-SHA-1-80
      - CMAC-AES-128
      - CMAC-AES-192
      - CMAC-AES-256
  - Symmetric Encryption Schemes
    - Supported:
      - AES-128-CBC
      - AES-192-CBC
      - AES-256-CBC
      - AES-128-CTR
      - AES-192-CTR
      - AES-256-CTR
    - Not supported:
      - 3-key TDES in CBC mode
      - XOR encryption scheme

In addition, the following options have been chosen:
  - Elliptical curve points are represented in compressed form.
  - Benign malleability is prevented by including the ephemeral public key as an input to the KDF (sec1-v2 p97).

## Supported platforms

Ruby 2.0 and above.

## Contributing

Bug reports and pull requests welcome! I happily accept any feedback that can improve this library's security.

## Disclaimer

While I have taken every effort to make this library as secure as possible, it is still an early version and has not yet been reviewed by a wide audience. Use at your own risk.

## Documentation

For complete documentation, see the [ECIES page on RubyDoc.info](http://rubydoc.info/gems/ecies).


CHANGELOG.md
Change log
====

This gem follows [Semantic Versioning 2.0.0](http://semver.org/spec/v2.0.0.html).
All classes and public methods are part of the public API.

0.4.0
---
Released on 2023-06-20

- Add support for OpenSSL 3.0.
- Remove `Crypt.private_key_from_hex` method.
  - Users can still utilize `OpenSSL::PKey::EC.new` to construct a PKey.
- `Crypt.public_key_from_hex` now raises `ArgumentError` on an invalid key, rather than an `OpenSSL::PKey::EC::Point::Error`.
- Explicitly require 'stringio' (thanks thekuwayama!).

0.3.0
---
Released on 2018-04-22

- Prevent benign malleability, as suggested in sec1-v2 page 97.
  - The ECIES process is modified to prevent benign malleability by including the ephemeral public key as an input to the KDF.
  - All encrypted output generated with previous versions cannot be decrypted with this version, and older versions cannot decrypt output generated with this version.
  - The choice was made to simply break compatibility early in this library's life, rather than add an extra configuration parameter that should almost always be unused.
- Add `Crypt.public_key_from_hex` method.
- Add `Crypt.private_key_from_hex` method.
- Remove support for hex-encoded keys in `Crypt#encrypt` and `Crypt#decrypt` methods. The above `*_from_hex` helper methods can be used instead.
- Remove `ec_group` option from `Crypt` constructor.
- Add `Crypt#to_s` method.

0.2.0
---
Released on 2018-04-19

- Add support for hex-encoded keys in `Crypt#encrypt` and `Crypt#decrypt` methods.
- Add new option `ec_group` in `Crypt` constructor.

0.1.0
----
Released on 2018-04-18

All core functionality is implemented:

- `ECIES::Crypt` class:
  - `encrypt` method
  - `decrypt` method
  - `kdf` method
  - `DIGESTS` constant
  - `CIPHERS` constant
  - `IV` constant


spec/crypt_spec.rb
require 'spec_helper'

describe ECIES::Crypt do

  describe 'Encryption and decryption' do

    it 'Encrypts and decrypts' do
      key = OpenSSL::PKey::EC.generate('secp256k1')
      crypt = ECIES::Crypt.new

      encrypted = crypt.encrypt(key, 'secret')
      expect(crypt.decrypt(key, encrypted)).to eq 'secret'

      expect{ ECIES::Crypt.new(mac_length: :full).decrypt(key, encrypted) }.to raise_error(OpenSSL::PKey::ECError)
      expect{ ECIES::Crypt.new(mac_digest: 'sha512').decrypt(key, encrypted) }.to raise_error(OpenSSL::PKey::ECError)
    end

    it 'Supports hex-encoded keys' do
      key = OpenSSL::PKey::EC.generate('secp256k1')
      public_key_hex = key.public_key.to_bn.to_s(16)

      public_key = ECIES::Crypt.public_key_from_hex(public_key_hex)

      expect(public_key.public_key).to eq key.public_key

      expect{ ECIES::Crypt.public_key_from_hex(public_key_hex, 'secp224k1') }.to raise_error(ArgumentError)
    end

    it 'Supports other EC curves' do
      key = OpenSSL::PKey::EC.generate('secp224k1')
      crypt = ECIES::Crypt.new

      encrypted = crypt.encrypt(key, 'secret')
      expect(crypt.decrypt(key, encrypted)).to eq 'secret'
    end

    context 'known value' do
      before(:all) do
        OpenSSL::PKey::EC.instance_eval do
          # Overwrites `generate` for both the key generated below, and the
          # ephemeral_key generated in the `encrypt` method.
          # The private_key is equal to '2', and the group is 'secp256k1'.
          def generate(group)
            OpenSSL::PKey::EC.new("0t\x02\x01\x01\x04 \x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x02\xA0\a\x06\x05+\x81\x04\x00\n\xA1D\x03B\x00\x04\xC6\x04\x7F\x94A\xED}m0E@n\x95\xC0|\xD8\\w\x8EK\x8C\xEF<\xA7\xAB\xAC\t\xB9\\p\x9E\xE5\x1A\xE1h\xFE\xA6=\xC39\xA3\xC5\x84\x19Fl\xEA\xEE\xF7\xF62e2f\xD0\xE1#d1\xA9P\xCF\xE5*")
          end
        end

        @key = OpenSSL::PKey::EC.generate('secp256k1')
      end

      [
        [ECIES::Crypt.new, "\x02\xC6\x04\x7F\x94A\xED}m0E@n\x95\xC0|\xD8\\w\x8EK\x8C\xEF<\xA7\xAB\xAC\t\xB9\\p\x9E\xE5C\x9E\xE0\x0FYBZ\xBB\xC8\x95\x93\xC1@\xC6+\xE2/yb\x065\xFF".b],
        [ECIES::Crypt.new(mac_length: :full), "\x02\xC6\x04\x7F\x94A\xED}m0E@n\x95\xC0|\xD8\\w\x8EK\x8C\xEF<\xA7\xAB\xAC\t\xB9\\p\x9E\xE5C\x9E\xE0\x0FYB\x03.\x1E\x92,[\rI\xBC\xCC\xFD%\xCD)9\v!]]A\xE0\xADc\xBA[\xA4\xF2\xB1\xB5\xC5)\xA4".b],
        [ECIES::Crypt.new(digest: 'sha512', mac_length: :full), "\x02\xC6\x04\x7F\x94A\xED}m0E@n\x95\xC0|\xD8\\w\x8EK\x8C\xEF<\xA7\xAB\xAC\t\xB9\\p\x9E\xE5\xA2Y\x1A\x7F\xB3\xB2\xA7\xDE\x03\xF4\xA6\e\xD1\x9F\xF9\xD5P\x06\x91\x8EiW\xC82\xD9\xBB\xD2\xC92\xE2\x9F\x15F.\x8C]\xE3Y2\xD3L\xE8\xC4\x9F\xBF\xA5S\x98\x9AYy_Y\xF8\x05\xE7\x19\x9E\xDA\vn;Bvm\xA2`i5:".b],
        [ECIES::Crypt.new(cipher: 'aes-256-cbc', mac_length: :full), "\x02\xC6\x04\x7F\x94A\xED}m0E@n\x95\xC0|\xD8\\w\x8EK\x8C\xEF<\xA7\xAB\xAC\t\xB9\\p\x9E\xE5\xDF\xCD\x95\xAD!m\xAA/Xv\"\x97\x04\xEE\x9F\xEB^\x1F\xA7\xC9n\xE3\x94l;\xBA\xF2\xBE\xCD\x83\x02+\x02\x9D\x18\x11\x9A\xBEz_\x8A\xDB\xA3\x00\xF7\x8A\x94G".b],
        [ECIES::Crypt.new(mac_digest: 'sha256', kdf_digest: 'sha512'), "\x02\xC6\x04\x7F\x94A\xED}m0E@n\x95\xC0|\xD8\\w\x8EK\x8C\xEF<\xA7\xAB\xAC\t\xB9\\p\x9E\xE5\xA2Y\x1A\x7F\xB3\xB2l\x9E|\xC4\xBCE r\xA6\xB1k\x93W\xE5d\xE4".b],
      ].each do |crypt, expected_value|
        it "matches for #{crypt.to_s}" do
          encrypted = crypt.encrypt(@key, 'secret')
          expect(encrypted).to eq expected_value
          expect(crypt.decrypt(@key, encrypted)).to eq 'secret'
        end
      end
    end

    it '#to_s' do
      [
        [ECIES::Crypt.new,                                             "KDF-SHA256_HMAC-SHA-256-128_AES-256-CTR"],
        [ECIES::Crypt.new(mac_length: :full),                          "KDF-SHA256_HMAC-SHA-256-256_AES-256-CTR"],
        [ECIES::Crypt.new(digest: 'sha512'),                           "KDF-SHA512_HMAC-SHA-512-256_AES-256-CTR"],
        [ECIES::Crypt.new(mac_digest: 'sha512'),                       "KDF-SHA256_HMAC-SHA-512-256_AES-256-CTR"],
        [ECIES::Crypt.new(mac_digest: 'sha512', kdf_digest: 'sha224'), "KDF-SHA224_HMAC-SHA-512-256_AES-256-CTR"],
        [ECIES::Crypt.new(cipher: 'aes-128-cbc'),                      "KDF-SHA256_HMAC-SHA-256-128_AES-128-CBC"],
      ].each do |crypt, expected_value|
        expect(crypt.to_s).to eq expected_value
      end
    end

    it 'Raises on unknown cipher or digest' do
      key = OpenSSL::PKey::EC.generate('secp256k1')

      expect{ ECIES::Crypt.new(digest: 'foo') }.to raise_error(RuntimeError)
      expect{ ECIES::Crypt.new(digest: 'md5') }.to raise_error(RuntimeError)
      expect{ ECIES::Crypt.new(cipher: 'foo') }.to raise_error(RuntimeError)
      expect{ ECIES::Crypt.new(cipher: 'aes-256-gcm') }.to raise_error(RuntimeError)
    end

    it 'Raises when key is missing' do
      key = OpenSSL::PKey::EC.new

      expect{ ECIES::Crypt.new.encrypt(key, 'secret') }.to raise_error(RuntimeError)
      expect{ ECIES::Crypt.new.decrypt(key, 'secret') }.to raise_error(RuntimeError)
    end
  end

  describe '#kdf' do
    it 'derives keys correctly' do
      sha256_test_vectors = [
        # [shared_secret, shared_info, expected_key]
        ['96c05619d56c328ab95fe84b18264b08725b85e33fd34f08', '', '443024c3dae66b95e6f5670601558f71'],
        ['96f600b73ad6ac5629577eced51743dd2c24c21b1ac83ee4', '', 'b6295162a7804f5667ba9070f82fa522'],
        ['22518b10e70f2a3f243810ae3254139efbee04aa57c7af7d', '75eef81aa3041e33b80971203d2c0c52', 'c498af77161cc59f2962b9a713e2b215152d139766ce34a776df11866a69bf2e52a13d9c7c6fc878c50c5ea0bc7b00e0da2447cfd874f6cf92f30d0097111485500c90c3af8b487872d04685d14c8d1dc8d7fa08beb0ce0ababc11f0bd496269142d43525a78e5bc79a17f59676a5706dc54d54d4d1f0bd7e386128ec26afc21'],
        ['7e335afa4b31d772c0635c7b0e06f26fcd781df947d2990a', 'd65a4812733f8cdbcdfb4b2f4c191d87', 'c0bd9e38a8f9de14c2acd35b2f3410c6988cf02400543631e0d6a4c1d030365acbf398115e51aaddebdc9590664210f9aa9fed770d4c57edeafa0b8c14f93300865251218c262d63dadc47dfa0e0284826793985137e0a544ec80abf2fdf5ab90bdaea66204012efe34971dc431d625cd9a329b8217cc8fd0d9f02b13f2f6b0b'],
      ]

      sha256_test_vectors.each do |shared_secret, shared_info, expected_key|
        shared_secret = [shared_secret].pack('H*')
        shared_info = [shared_info].pack('H*')
        expected_key = [expected_key].pack('H*')

        computed_key = ECIES::Crypt.new(kdf_shared_info: shared_info).kdf(shared_secret, expected_key.size, '')
        expect(computed_key).to eq expected_key
      end
    end

    it 'concats kdf_shared_info with shared_info_suffix' do
      shared_secret = ['22518b10e70f2a3f243810ae3254139efbee04aa57c7af7d'].pack('H*')
      shared_info = ['75eef81aa3041e33'].pack('H*')
      shared_info_suffix = ['b80971203d2c0c52'].pack('H*')
      expected_key = ['c498af77161cc59f2962b9a713e2b215152d139766ce34a776df11866a69bf2e52a13d9c7c6fc878c50c5ea0bc7b00e0da2447cfd874f6cf92f30d0097111485500c90c3af8b487872d04685d14c8d1dc8d7fa08beb0ce0ababc11f0bd496269142d43525a78e5bc79a17f59676a5706dc54d54d4d1f0bd7e386128ec26afc21'].pack('H*')

      computed_key = ECIES::Crypt.new(kdf_shared_info: shared_info).kdf(shared_secret, expected_key.size, shared_info_suffix)
      expect(computed_key).to eq expected_key
    end

    it 'raises when size is invalid' do
      expect{ ECIES::Crypt.new.kdf('a', -1, '') }.to raise_error(RuntimeError)
      expect{ ECIES::Crypt.new.kdf('a', 32 * 2**32, '') }.to raise_error(RuntimeError)
    end
  end
end


spec/spec_helper.rb
if ENV['COVERAGE']
  require 'simplecov'
  SimpleCov.start
end

require 'ecies'


lib/ecies/crypt.rb
module ECIES
  # Provides functionality for encrypting and decrypting messages using ECIES.
  # Encapsulates the configuration parameters chosen for ECIES.
  class Crypt

    # The allowed digest algorithms for ECIES.
    DIGESTS = %w{SHA224 SHA256 SHA384 SHA512}

    # The allowed cipher algorithms for ECIES.
    CIPHERS = %w{AES-128-CBC AES-192-CBC AES-256-CBC AES-128-CTR AES-192-CTR AES-256-CTR}

    # The initialization vector used in ECIES. Quoting from sec1-v2:
    # "When using ECIES, some exception are made. For the CBC and CTR modes, the
    # initial value or initial counter are set to be zero and are omitted from
    # the ciphertext. In general this practice is not advisable, but in the case
    # of ECIES it is acceptable because the definition of ECIES implies the
    # symmetric block cipher key is only to be used once.
    IV = ("\x00" * 16).force_encoding(Encoding::BINARY)

    # Creates a new instance of {Crypt}.
    #
    # @param cipher [String] The cipher algorithm to use. Must be one of
    #     {CIPHERS}.
    # @param digest [String,OpenSSL::Digest] The digest algorithm to use for
    #     HMAC and KDF. Must be one of {DIGESTS}.
    # @param mac_length [:half,:full] The length of the mac. If :half, the mac
    #     length will be equal to half the mac_digest's digest_legnth. If
    #     :full, the mac length will be equal to the mac_digest's
    #     digest_length.
    # @param kdf_digest [String,OpenSSL::Digest,nil] The digest algorithm to
    #     use for KDF. If not specified, the `digest` argument will be used.
    # @param mac_digest [String,OpenSSL::Digest,nil] The digest algorithm to
    #     use for HMAC. If not specified, the `digest` argument will be used.
    # @param kdf_shared_info [String] Optional. A string containing the shared
    #     info used for KDF, also known as SharedInfo1.
    # @param mac_shared_info [String] Optional. A string containing the shared
    #     info used for MAC, also known as SharedInfo2.
    def initialize(cipher: 'AES-256-CTR', digest: 'SHA256', mac_length: :half, kdf_digest: nil, mac_digest: nil, kdf_shared_info: '', mac_shared_info: '')
      @cipher = OpenSSL::Cipher.new(cipher)
      @mac_digest = OpenSSL::Digest.new(mac_digest || digest)
      @kdf_digest = OpenSSL::Digest.new(kdf_digest || digest)
      @kdf_shared_info = kdf_shared_info
      @mac_shared_info = mac_shared_info

      CIPHERS.include?(@cipher.name) or raise "Cipher must be one of #{CIPHERS}"
      DIGESTS.include?(@mac_digest.name) or raise "Digest must be one of #{DIGESTS}"
      DIGESTS.include?(@kdf_digest.name) or raise "Digest must be one of #{DIGESTS}"
      [:half, :full].include?(mac_length) or raise "mac_length must be :half or :full"

      @mac_length = @mac_digest.digest_length
      @mac_length /= 2 if mac_length == :half
    end

    # Encrypts a message to a public key using ECIES.
    #
    # @param key [OpenSSL::EC:PKey] The public key.
    # @param message [String] The plain-text message.
    # @return [String] The octet string of the encrypted message.
    def encrypt(key, message)
      key.public_key? or raise "Must have public key to encrypt"
      @cipher.reset

      ephemeral_key = OpenSSL::PKey::EC.generate(key.group)
      ephemeral_public_key_octet = ephemeral_key.public_key.to_octet_string(:compressed)

      shared_secret = ephemeral_key.dh_compute_key(key.public_key)

      key_pair = kdf(shared_secret, @cipher.key_len + @mac_length, ephemeral_public_key_octet)
      cipher_key = key_pair.byteslice(0, @cipher.key_len)
      hmac_key = key_pair.byteslice(-@mac_length, @mac_length)

      @cipher.encrypt
      @cipher.iv = IV
      @cipher.key = cipher_key
      ciphertext = @cipher.update(message) + @cipher.final

      mac = OpenSSL::HMAC.digest(@mac_digest, hmac_key, ciphertext + @mac_shared_info).byteslice(0, @mac_length)

      ephemeral_public_key_octet + ciphertext + mac
    end

    # Decrypts a message with a private key using ECIES.
    #
    # @param key [OpenSSL::EC:PKey] The private key.
    # @param encrypted_message [String] Octet string of the encrypted message.
    # @return [String] The plain-text message.
    def decrypt(key, encrypted_message)
      key.private_key? or raise "Must have private key to decrypt"
      @cipher.reset

      ephemeral_public_key_length = key.group.generator.to_octet_string(:compressed).bytesize
      ciphertext_length = encrypted_message.bytesize - ephemeral_public_key_length - @mac_length
      ciphertext_length > 0 or raise OpenSSL::PKey::ECError, "Encrypted message too short"

      ephemeral_public_key_octet = encrypted_message.byteslice(0, ephemeral_public_key_length)
      ciphertext = encrypted_message.byteslice(ephemeral_public_key_length, ciphertext_length)
      mac = encrypted_message.byteslice(-@mac_length, @mac_length)

      ephemeral_public_key = OpenSSL::PKey::EC::Point.new(key.group, OpenSSL::BN.new(ephemeral_public_key_octet, 2))

      shared_secret = key.dh_compute_key(ephemeral_public_key)

      key_pair = kdf(shared_secret, @cipher.key_len + @mac_length, ephemeral_public_key_octet)
      cipher_key = key_pair.byteslice(0, @cipher.key_len)
      hmac_key = key_pair.byteslice(-@mac_length, @mac_length)

      computed_mac = OpenSSL::HMAC.digest(@mac_digest, hmac_key, ciphertext + @mac_shared_info).byteslice(0, @mac_length)
      computed_mac == mac or raise OpenSSL::PKey::ECError, "Invalid Message Authenticaton Code"

      @cipher.decrypt
      @cipher.iv = IV
      @cipher.key = cipher_key

      @cipher.update(ciphertext) + @cipher.final
    end

    # Key-derivation function, compatible with ANSI-X9.63-KDF
    #
    # @param shared_secret [String] The shared secret from which the key will
    #     be derived.
    # @param length [Integer] The length of the key to generate.
    # @param shared_info_suffix [String] The suffix to append to the
    #     shared_info.
    # @return [String] Octet string of the derived key.
    def kdf(shared_secret, length, shared_info_suffix)
      length >=0 or raise "length cannot be negative"
      return "" if length == 0

      if length / @kdf_digest.digest_length >= 0xFF_FF_FF_FF
        raise "length too large"
      end

      io = StringIO.new(String.new)
      counter = 0

      loop do
        counter += 1
        counter_bytes = [counter].pack('N')

        io << @kdf_digest.digest(shared_secret + counter_bytes + @kdf_shared_info + shared_info_suffix)
        if io.pos >= length
          return io.string.byteslice(0, length)
        end
      end
    end

    # @return [String] A string representing this Crypt's parameters.
    def to_s
      "KDF-#{@kdf_digest.name}_" +
      "HMAC-SHA-#{@mac_digest.digest_length * 8}-#{@mac_length * 8}_" +
      @cipher.name
    end

    # Converts a hex-encoded public key to an `OpenSSL::PKey::EC`.
    #
    # @param hex_string [String] The hex-encoded public key.
    # @param ec_group [OpenSSL::PKey::EC::Group,String] The elliptical curve
    #     group for this public key.
    # @return [OpenSSL::PKey::EC] The public key.
    # @raise [ArgumentError] If the public key is invalid.
    def self.public_key_from_hex(hex_string, ec_group = 'secp256k1')
      ec_group = OpenSSL::PKey::EC::Group.new(ec_group) if ec_group.is_a?(String)

      sequence = OpenSSL::ASN1.Sequence([
        OpenSSL::ASN1.Sequence([
          OpenSSL::ASN1.ObjectId("id-ecPublicKey"),
          OpenSSL::ASN1::ObjectId(ec_group.curve_name),
        ]),
        OpenSSL::ASN1.BitString([hex_string].pack('H*')),
      ])

      OpenSSL::PKey::EC.new(sequence.to_der)
    end
  end
end
